posts/boost-your-databricks-jobs-and-queries-performance.md

```diff
@@ -0,0 +1,66 @@
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-boost-performance-databricks/images/2023/02/databricks-autoscaling.png)
+
+Don't get me wrong, this is the whole point of autoscaling but the wait time between each upsizing events has an impact. A streaming job will have more data to process by the time the first upzise is finished and then will probably have to upsize again as new data is processed. If your cluster has been idle with a low number of workers - running a heavy command will take longer because it will have to first complete the upsize before starting that command.
+
+One last detail is due to node termination, any data cached will be lost when the cluster completes its downsizing. If cost saving is of a higher priority than perfomance - autoscaling is definetly the way to go! Be sure to read Databricks [best practices about autoscaling](https://docs.databricks.com/clusters/cluster-config-best-practices.html#autoscaling).
+
+# Small cluster with larger VMs is better for complex ETL
+
+A complex ETL job that requires unions and joins across multiple tables will run faster on a cluster that uses very large instance types. Having a reduced number of workers will minimize the amount of data shuffled. Your large workers will also have more RAM available and have less `OUT_OF_MEMORY` errors like below (DLT pipeline):
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-boost-performance-databricks/images/2023/02/databricks-out-of-memory.png)
+
+ A low number of large instance types is always better than a high number of small instance types, for more details, I recommend reading [the Databricks page about cluster size consideration](https://docs.databricks.com/clusters/cluster-config-best-practices.html#cluster-sizing-considerations).
+
+One risk to consider is spot termination of a large worker will have way more of an impact on your cluster's resources than losing a small worker.
+
+# Optimize and Z-ordering are your friends
+
+The [Optimize command](https://docs.databricks.com/delta/optimize.html) is very useful and will consolidate small files into larger ones to improve your query speed. You can even select a subset of your data by inputing a `where` in the `optimize` command. Databricks recommends running `optimize` every night when the EC2 spot prices are low.
+
+Z-ordering can be run on columns that are often used in your query's predicates and that are high cardinality. For example, if you have a table with daily stock price, you will want to z-order on the stock ticker column because there is a high number of tickers. Ticker will often be in your queries as it is a common predicate. Partitionning on the ticker column is not recommended and will create too many small partitions. I recommend reading [the Databricks page](https://docs.databricks.com/delta/data-skipping.html#what-is-z-ordering) about z-ordering for more information.
```

Comment:
```suggestion
Z-ordering can be run on columns that are often used in your query's predicates and are high cardinality. For example, if you have a table with daily stock prices, you will want to z-order on the stock ticker column because there is a high number of tickers. Ticker will often be in your queries as it is a common predicate. Partitioning on the ticker column is not recommended and will create too many small partitions. I recommend reading [this Databricks page](https://docs.databricks.com/delta/data-skipping.html#what-is-z-ordering) about z-ordering for more information.
```